mechanicalsoup/stateful_browser.py

```diff
@@ -298,6 +298,34 @@ def follow_link(self, link=None, *args, **kwargs):
 
         return self.open_relative(link['href'], **request_kwargs)
 
+    def download_link(self, filename, link=None, *args, **kwargs):
+        """Downloads the contents of a link to a file. This function behaves
+        similarly to :func:`follow_link`, but the browser state will
+        not change when calling this function.
+
+        :param filename: Filesystem path where the page contents will be
+            downloaded. If the file already exists, it will be overwritten.
+
+        Other arguments are the same as :func:`follow_link` (``link``
+        can either be a bs4.element.Tag or a URL regex, other
+        arguments are forwarded to :func:`find_link`).
+
+        :return: `requests.Response
+            <http://docs.python-requests.org/en/master/api/#requests.Response>`__
+            object.
+        """
+        link = self._find_link_internal(link, args, kwargs)
+        url = self.absolute_url(link['href'])
+        response = self.session.get(url)
+        if self.raise_on_404 and response.status_code == 404:
+            raise LinkNotFoundError()
+
+        # Save the response content to file
```

Comment:
Dear @moy, thanks for your work on this!

May i humbly suggest to make writing to the filesystem optional? My proposal would be to change the method signature to:
`def download_link(self, link=None, filename=None, *args, **kwargs)`

... and then conditionally run the code like:
```
if filename is not None:
    with open(filename, 'wb') as f:
        f.write(response.content)
```

----

A further improvement could be to accept an already opened file handle as value for the `filename` argument like
```
if filename is not None:
    if isinstance(filename, six.string_types):
        with open(filename, 'wb') as f:
            f.write(response.content)
    elif hasattr(filename, 'write'):
        filename.write(response.content)
```

When doing this, one would be able to also pass an in-memory string buffer like an instance of `StringIO` to the `download_link` method.

Reply:
Thanks!